xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java

```diff
@@ -0,0 +1,343 @@
+                                                   .addService(server.getEndpointDiscoveryServiceImpl())
+                                                   .addService(server.getListenerDiscoveryServiceImpl())
+                                                   .addService(server.getRouteDiscoveryServiceImpl())
+                                                   .addService(server.getSecretDiscoveryServiceImpl())
+                                                   .addService(server.getAggregatedDiscoveryServiceImpl())
+                                                   .useBlockingTaskExecutor(true)
+                                                   .build();
+        sb.route().requestTimeoutMillis(0).build(grpcService);
+    }
+
+    private void watchRepository(Repository repository, String fileName, Revision revision,
+                                 BiFunction<Collection<Entry<?>>, Revision, Boolean> snapshotConverter) {
+        final CompletableFuture<Revision> future = repository.watch(revision, "/**");
+        future.handleAsync((BiFunction<Revision, Throwable, Void>) (watchedRevision, cause) -> {
+            if (stop) {
+                return null;
+            }
+            if (cause != null) {
+                logger.warn("Unexpected exception is raised while watching {}. Try watching after {} seconds..",
+                            repository, BACKOFF_SECONDS, cause);
+                CONTROL_PLANE_EXECUTOR.schedule(
+                        () -> watchRepository(repository, fileName, revision, snapshotConverter),
+                        BACKOFF_SECONDS, TimeUnit.SECONDS);
+                return null;
+            }
+            final CompletableFuture<Map<String, Entry<?>>> entriesFuture =
+                    repository.find(watchedRevision, "/**/" + fileName);
+            entriesFuture.handleAsync(
+                    (BiFunction<Map<String, Entry<?>>, Throwable, Void>) (entries, findCause) -> {
+                        if (stop) {
+                            return null;
+                        }
+                        if (findCause != null) {
+                            logger.warn(
+                                    "Unexpected exception is raised while finding {}" +
+                                    " with revision {} from {}. Try watching after {} seconds..",
+                                    fileName, watchedRevision, repository, BACKOFF_SECONDS, findCause);
+                            CONTROL_PLANE_EXECUTOR.schedule(
+                                    () -> watchRepository(repository, fileName,
+                                                          watchedRevision, snapshotConverter),
+                                    BACKOFF_SECONDS, TimeUnit.SECONDS);
+                            return null;
+                        }
+                        final Boolean converted = snapshotConverter.apply(entries.values(), watchedRevision);
```

Comment:
`snapshotConverter` sounds like it returns a converted value.
How about renaming it to `rawResourceListener`?

Reply:
How about renaming it to `creatingSnapshotFunction` since it generates a new snapshot when a resource is updated?